gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (C) 2020 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.gc.base;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.OptionalInt;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.spark.sql.SparkSession;
+import org.projectnessie.api.params.FetchOption;
+import org.projectnessie.client.StreamingUtil;
+import org.projectnessie.client.api.NessieApiV1;
+import org.projectnessie.error.NessieNotFoundException;
+import org.projectnessie.model.Branch;
+import org.projectnessie.model.RefLogResponse;
+import org.projectnessie.model.Reference;
+import org.projectnessie.model.ReferenceMetadata;
+import org.projectnessie.model.Tag;
+
+/**
+ * Encapsulates the logic to retrieve all (both live and expired) content including their content
+ * keys over all commits in all named-references.
+ */
+public class GCImpl {
+  private final GCParams gcParams;
+
+  /**
+   * Instantiates a new Gc.
+   *
+   * @param configuration the configuration
+   * @param defaultCutOffTimeStamp the default cut off time stamp
+   * @param deadReferenceCutOffTimeStamp optional cut off time for dead reference
+   * @param cutOffTimestampPerRef the cutoff timestamp per ref
+   */
+  public GCImpl(
+      Map<String, String> configuration,
```

Comment:
Should take only GCParams as a parameter.
The validation is just unnecessary here w/ immutables,